crucible/test/helpers/Main.hs

```diff
@@ -1,20 +1,84 @@
 {-# LANGUAGE AllowAmbiguousTypes #-}
 module Main where
 
+import Control.Lens ((^.))
 import Data.List (isInfixOf)
+import Data.Maybe (fromMaybe)
 
 import Test.Hspec
 import Test.Tasty
 import Test.Tasty.Hspec (testSpec)
+import qualified Test.Tasty.HUnit as TTH
 
+import Data.Parameterized.Nonce (newIONonceGenerator)
+import Data.Parameterized.Some (Some(Some))
+import Lang.Crucible.Backend (SomeBackend(..))
+import qualified Lang.Crucible.Backend as LCB
+import Lang.Crucible.Backend.Simple (newSimpleBackend)
 import Lang.Crucible.Panic
+import Lang.Crucible.Simulator.SimError (SimErrorReason(GenericSimError))
+import What4.Expr (EmptyExprBuilderState(..))
+import What4.Expr.Builder (newExprBuilder)
+import qualified What4.Interface as W4I
+import What4.FloatMode (FloatModeRepr(FloatIEEERepr))
+import qualified What4.LabeledPred as W4L
+import qualified What4.ProgramLoc as W4P
 
 import qualified Panic as P
 
 main :: IO ()
 main =
   defaultMain =<< panicTests
 
+mkBackend :: IO (Some SomeBackend)
+mkBackend = do
+  Some nonceGen <- newIONonceGenerator
+  sym <- newExprBuilder FloatIEEERepr EmptyExprBuilderState nonceGen
+  Some . SomeBackend <$> newSimpleBackend sym
+
+assumePred :: LCB.IsSymBackend sym bak => bak -> W4I.Pred sym -> IO ()
+assumePred bak p =
+  LCB.addAssumption bak (LCB.GenericAssumption W4P.initializationLoc "" p)
+
+backendTests :: TestTree
+backendTests =
+  testGroup
+  "Backend"
+  [ TTH.testCase "push/pop nothing" $ do
+      Some (SomeBackend bak) <- mkBackend
+      asmps <- LCB.popAssumptionFrame bak =<< LCB.pushAssumptionFrame bak
+      p <- LCB.assumptionsPred (LCB.backendGetSym bak) asmps
+      Just True TTH.@=? W4I.asConstantPred p
+  , TTH.testCase "push/pop assumptions" $ do
```

Comment:
Could you leave a high-level comment next to each test case that describes the property it is testing for? I'll admit that the code is dense enough that it's hard to tell just by looking at it...

Reply:
I'm actually a bit surprised to hear that you found the code challenging to read! It says:

1. Make a fresh predicate `a`
2. Assume `a`
3. Push a frame
4. Make a fresh predicate `b`
5. Assume `b`
6. Pop the frame
7. Assert that the popped assumptions are equal to `b`

To me, that's a pretty direct reading (basically one step per line) of lines 55-62.

I will add a comment that says "check that popping returns assumptions from the popped frame".